lib/srv/db/postgres/users.go

```diff
@@ -81,8 +87,179 @@ func (e *Engine) ActivateUser(ctx context.Context, sessionCtx *common.Session) e
+	if err != nil {
+		e.Log.WithError(err).Warn("Failed to apply permissions.")
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+type TablePermission struct {
+	Privilege string `json:"privilege"`
+	Schema    string `json:"schema"`
+	Table     string `json:"table"`
+}
+
+type Permissions struct {
+	Tables []TablePermission `json:"tables"`
+}
+
+var pgTablePerms = map[string]struct{}{
+	"SELECT":     {},
+	"INSERT":     {},
+	"UPDATE":     {},
+	"DELETE":     {},
+	"TRUNCATE":   {},
+	"REFERENCES": {},
+	"TRIGGER":    {},
+}
+
+func checkPgPermission(objKind, perm string) error {
+	// for now, only tables are supported. ignore other kinds of objects.
+	if objKind != permissions.ObjectKindTable {
+		return nil
+	}
+
+	normalized := strings.ToUpper(strings.TrimSpace(perm))
+	_, found := pgTablePerms[normalized]
+	if !found {
+		return trace.BadParameter("unrecognized %q Postgres permission: %q", objKind, perm)
+	}
+	return nil
+}
+
+// convertPermissions converts the permissions into a stable format expected by the stored procedure. It also filters out any unsupported objects.
+func convertPermissions(perms permissions.PermissionSet) (*Permissions, error) {
+	var out Permissions
+	var errors []error
+	for permission, objects := range perms {
+		for _, obj := range objects {
+			if err := checkPgPermission(obj.GetSpec().ObjectKind, permission); err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			if obj.GetSpec().ObjectKind == permissions.ObjectKindTable {
+				out.Tables = append(out.Tables, TablePermission{
+					Privilege: permission,
+					Schema:    obj.GetSpec().Schema,
+					Table:     obj.GetSpec().Name,
+				})
+			}
+		}
+	}
+	if len(errors) > 0 {
+		return nil, trace.NewAggregate(errors...)
+	}
+	return &out, nil
+}
+
+func (e *Engine) granularPermissionsEnabled(sessionCtx *common.Session) bool {
+	allow, _, _ := sessionCtx.Checker.GetDatabasePermissions(sessionCtx.Database)
+	if len(allow) == 0 {
+		return false
+	}
+
+	if len(sessionCtx.DatabaseRoles) > 0 {
+		return false
+	}
+
+	return true
+}
+
+func (e *Engine) applyPermissions(ctx context.Context, sessionCtx *common.Session) error {
+	allow, _, err := sessionCtx.Checker.GetDatabasePermissions(sessionCtx.Database)
+	if err != nil {
+		e.Log.WithError(err).Error("Failed to calculate effective database permissions.")
+		return trace.Wrap(err)
+	}
+	if len(allow) == 0 {
+		e.Log.Infof("Skipping applying fine-grained permissions: none to apply.")
+		return nil
+	}
+
+	if len(sessionCtx.DatabaseRoles) > 0 {
+		e.Log.Errorf("Cannot apply fine-grained permissions: non-empty list of database roles (%v).", sessionCtx.DatabaseRoles)
+		return trace.BadParameter("fine-grained database permissions and database roles are mutually exclusive, yet both were provided.")
+	}
+
+	rules, err := e.AuthClient.GetDatabaseObjectImportRules(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	conn, err := e.connectAsAdmin(ctx, sessionCtx, false)
```

Comment:
Will this work when a default database is set? All procedures should be stored in the "default" database.

Reply:
No, these must be created in the session database; I'll add a fix. 
> All procedures should be stored in the "default" database.
This works for user management, because users are global for a given database server, but it doesn't work for permission management: these are per-database, so the procedure must be created in the session database. As an alternative, I can change the approach away from the stored procedures, or I can create the stored procedure in `pg_temp` schema each time it is needed. I'll try to go with `pg_temp` solution for now.

Reply:
Hmm i couldn't find official docs on `pg_temp`. How long does it last? like will the procedures still be there for deactivation/removal? I will try it out later today.

Reply:
It is supposed to last only for the current session, but I'm having trouble locating the relevant docs; just a ton of SO remarks.

Reply:
I just tested this and this one surprisingly working. What counts as a session? Don't we do four admin sessions (activate, update, deactivate, remove) during a `tsh db connect`? Anyway, we can resolve this now.

Reply:
I believe it is equivalent to a particular database connection, and third party comments support that, although I couldn't find a official definition with my brief search. In principle we could treat _all_ our stored procs as temporary, but maybe there is a reason to do otherwise?

Reply:
My bad. I see you install them on each connection. that explains it.